javascript/components/MapView.tsx

```diff
@@ -278,540 +257,603 @@ export interface MapViewState {
-      return [this.props.contentInset];
-    }
-
-    return this.props.contentInset;
-  }
-
-  _setNativeRef(nativeRef: RCTMGLMapViewRefType): void {
-    this._nativeRef = nativeRef;
-    super._runPendingNativeCommands(nativeRef);
-  }
-
-  setNativeProps(props: NativeProps): void {
-    if (this._nativeRef) {
-      this._nativeRef.setNativeProps(props);
-    }
-  }
-
-  _setStyleURL(props: MapViewProps): void {
-    // user set a styleURL, no need to alter props
-    if (props.styleURL) {
-      return;
-    }
-
-    // user set styleJSON pass it to styleURL
-    if (props.styleJSON && !props.styleURL) {
-      props.styleURL = props.styleJSON;
-    }
-
-    // user neither set styleJSON nor styleURL
-    // set defaultStyleUrl
-    if (!props.styleJSON || !props.styleURL) {
-      props.styleURL = defaultStyleURL;
-    }
-  }
-
-  render(): ReactElement {
-    const props = {
-      ...this.props,
-      contentInset: this._getContentInset(),
-      style: styles.matchParent,
-    };
-
-    this._setStyleURL(props);
-
-    const callbacks = {
-      ref: (ref: RCTMGLMapViewRefType) => this._setNativeRef(ref),
-      onPress: this._onPress,
-      onLongPress: this._onLongPress,
-      onMapChange: this._onChange,
-      onAndroidCallback: isAndroid() ? this._onAndroidCallback : undefined,
-    };
-
-    let mapView: ReactElement | null = null;
-    if (isAndroid() && !this.props.surfaceView && this.state.isReady) {
-      mapView = (
-        <RCTMGLAndroidTextureMapView {...props} {...callbacks}>
-          {this.props.children}
-        </RCTMGLAndroidTextureMapView>
-      );
-    } else if (this.state.isReady) {
-      mapView = (
-        <RCTMGLMapView {...props} {...callbacks}>
-          {this.props.children}
-        </RCTMGLMapView>
+
+        // user set styleJSON pass it to styleURL
+        if (props.styleJSON && !props.styleURL) {
+          props.styleURL = props.styleJSON;
+        }
+
+        // user neither set styleJSON nor styleURL
+        // set defaultStyleUrl
+        if (!props.styleJSON || !props.styleURL) {
+          props.styleURL = defaultStyleURL;
+        }
+      };
+
+      const nativeProps = useMemo(() => {
+        return {
+          ...props,
+          localizeLabels,
+          scrollEnabled,
+          pitchEnabled,
+          rotateEnabled,
+          attributionEnabled,
+          logoEnabled,
+          surfaceView,
+          regionWillChangeDebounceTime,
+          regionDidChangeDebounceTime,
+          contentInsetValue,
+          style: styles.matchParent,
+        };
+      }, [
+        localizeLabels,
+        scrollEnabled,
+        pitchEnabled,
+        rotateEnabled,
+        attributionEnabled,
+        logoEnabled,
+        surfaceView,
+        regionWillChangeDebounceTime,
+        regionDidChangeDebounceTime,
+        contentInsetValue,
+        props,
+        contentInsetValue,
+        styles.matchParent,
+      ]);
+
+      _setStyleURL(nativeProps);
```

Comment:
Do we need to run this on each re-render? Maybe we could use `useEffect` and put `props.styleJSON `, `props.styleURL` as deps.

Reply:
For the first draft I tried to keep every piece of logic the same as the old class based component, even though in cases like this it doesn't make sense. I just wanted  to make sure I don't break things. Will modify the code and add dependencies where needed in the final PR.

Reply:
For the first draft I tried to keep every piece of logic the same as the old class based component, even though in cases like this it doesn't make sense. I just wanted to make sure I don't break things. Will modify the code and add dependencies where needed in the final PR.